src/core/utils/power_calibration.hpp

```diff
@@ -0,0 +1,167 @@
+     *
+     */
+    Error SetChannelTargetPower(uint8_t aChannel, int16_t aTargetPower);
+
+    /**
+     * Get the raw power setting for the given channel.
+     *
+     * Platform radio layer should parse the raw power setting based on the radio layer defined format and set the
+     * parameters of each radio hardware module.
+     *
+     * @param[in]      aChannel                The radio channel.
+     * @param[out]     aRawPowerSetting        A pointer to the raw power setting byte array.
+     * @param[in,out]  aRawPowerSettingLength  On input, a pointer to the size of @p aRawPowerSetting.
+     *                                         On output, a pointer to the length of the raw power setting data.
+     *
+     * @retval  kErrorNone         Successfully got the target power.
+     * @retval  kErrorInvalidArgs  The @p aChannel is invalid or @p aRawPowerSetting is nullptr.
+     * @retval  kErrorNotFound     The raw power setting for the @p aChannel was not found.
+     *
+     */
+    Error GetRawPowerSetting(uint8_t aChannel, uint8_t *aRawPowerSetting, uint16_t *aRawPowerSettingLength);
+
+private:
+    class CalibratedPowerEntry
+    {
+    public:
+        static constexpr uint16_t kMaxRawPowerSettingSize = OPENTHREAD_CONFIG_RAW_POWER_SETTING_SIZE;
+
+        CalibratedPowerEntry(void)
+            : mActualPower(kInvalidPower)
+            , mLength(0)
+        {
+        }
+
+        void    Init(int16_t aActualPower, const uint8_t *aRawPowerSetting, uint16_t aRawPowerSettingLength);
+        Error   GetRawPowerSetting(uint8_t *aRawPowerSetting, uint16_t *aRawPowerSettingLength);
+        int16_t GetActualPower(void) const { return mActualPower; }
+        bool    Matches(int16_t aActualPower) const { return aActualPower == mActualPower; }
+
+    private:
+        int16_t  mActualPower;
+        uint8_t  mSettings[kMaxRawPowerSettingSize];
+        uint16_t mLength;
+    };
+
+    bool IsChannelValid(uint8_t aChannel) const
+    {
+        return ((aChannel >= Radio::kChannelMin) && (aChannel <= Radio::kChannelMax));
+    }
+
+    static constexpr uint8_t  kInvalidIndex           = NumericLimits<uint8_t>::kMax;
+    static constexpr uint16_t kInvalidPower           = NumericLimits<int16_t>::kMax;
+    static constexpr uint16_t kMaxNumCalibratedPowers = OPENTHREAD_CONFIG_NUM_CALIBRATED_POWER_ENTRIES;
+    static constexpr uint16_t kNumChannels            = Radio::kChannelMax - Radio::kChannelMin + 1;
+
+    static_assert(kMaxNumCalibratedPowers < NumericLimits<uint8_t>::kMax,
+                  "kMaxNumCalibratedPowers is larger than or equal to max");
+
+    typedef Array<CalibratedPowerEntry, kMaxNumCalibratedPowers> CalibratedPowerTable;
+
+    bool                 mPowerUpdated;
+    uint8_t              mChannel;
+    int16_t              mTargetPowerTable[kNumChannels];
+    uint8_t              mCalibratedPowerIndex;
+    CalibratedPowerTable mCalibratedPowerTables[kNumChannels];
```

Comment:
@zhanglongxia regarding combing the OT configs into one, I just noticed that there may be a downside to it:
- Platform can override the `weak` versions of `otPlatRadio{}` and provide its own
- But in that case, the `PowerCalibration` class is still compiled in and included in `Instance`
- So we end up using all the RAM space for member variable here (power table, etc) which are not really used
- So we can waste memory unnecessarily (may be fine for posix platforms, but not good for more resource limited platforms).

Maybe we should consider going back to two config model. Thoughts?

Reply:
Good catch. I have updated the code to add two configs.